fs/metrics.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "rocksdb/env.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+struct ZenFSMetrics {
+public:
+  ZenFSMetrics() {}
+  virtual ~ZenFSMetrics() {}
+public:
+  // Add a reporter named label.
+  // You can give a type for type-checking.
+  virtual void AddReporter(uint32_t label, uint32_t type = 0) = 0;
+  // Report a value for the reporter named label.
+  // You can give a type for type-checking.
+  virtual void Report(uint32_t label, size_t value, uint32_t type_check = 0) = 0;
+public: 
+  // Syntactic sugars for type-checking.
+  // Overwrite them if you think type-checking is necessary.
+  virtual void ReportQPS(uint32_t label, size_t qps) { Report(label, qps, 0); }
+  virtual void ReportThroughput(uint32_t label, size_t throughput) { Report(label, throughput, 0); }
+  virtual void ReportLatency(uint32_t label, size_t latency) { Report(label, latency, 0); }
+  virtual void ReportGeneral(uint32_t label, size_t data) { Report(label, data, 0); }
+  // and more
+};
+
+struct NoZenFSMetrics : public ZenFSMetrics {
+  NoZenFSMetrics() : ZenFSMetrics() {}
+  virtual ~NoZenFSMetrics() {}
+public:
+  virtual void AddReporter(uint32_t label, uint32_t type = 0) override {}
+  virtual void Report(uint32_t label, size_t value, uint32_t type_check = 0) override {}
+};
+
+// The implementation of this class will start timing when initialized, 
+// stop timing when it is destructured, 
+// and report the difference in time to the target label via metrics->ReportLatency().
+// By default, the method to collect the time will be to call env->NowMicros().
+struct ZenFSMetricsLatencyGuard {
+  std::shared_ptr<ZenFSMetrics> metrics_;
+  uint32_t label_;
+  Env* env_;
+  uint64_t begin_time_micro_;
+
+  ZenFSMetricsLatencyGuard(std::shared_ptr<ZenFSMetrics> metrics, uint32_t label, Env* env)
+  : metrics_(metrics), label_(label), env_(env), begin_time_micro_(GetTime()) {}
+
+  virtual ~ZenFSMetricsLatencyGuard() {
+    uint64_t end_time_micro_ = GetTime();
+    assert(end_time_micro_ > begin_time_micro_);
+    metrics_->ReportLatency(label_, Report(end_time_micro_ - begin_time_micro_)); 
+  }
+  // overwrite this function if you wish to capture time by other methods.
+  virtual uint64_t GetTime() { return env_->NowMicros(); }
+  // overwrite this function if you do not intend to report delays measured in microseconds.
+  virtual uint64_t Report(uint64_t time) { return time; }
+};
+
+// Names of Reporter that may be used for statistics.
+enum ZenFSMetricsHistograms : uint32_t {
+  ZENFS_HISTOGRAM_ENUM_MIN, 
+
+  ZENFS_FG_WRITE_LATENCY,
+  ZENFS_BG_WRITE_LATENCY,
+  
+  ZENFS_READ_LATENCY,
+  ZENFS_FG_SYNC_LATENCY,
+  ZENFS_BG_SYNC_LATENCY,
+  ZENFS_IO_ALLOC_WAL_LATENCY,
+  ZENFS_IO_ALLOC_NON_WAL_LATENCY,
```

Comment:
It seems that `WAL_LATENCY` and `NON_WAL_LATENCY` is not useful in this ZenFS. These are distinguished from each other only in TerarkDB's own fork of ZenFS.

Reply:
> 

@skyzh Noticed, thanks. I think we could make use of these flags soon. We are now moving from our own fork to the official ZenFS, so more patches are coming (don't worry about the waste of these flags, for now, LOL).